### ISSUE TITLE
bug: Plugin config values might be unnecessarily expanded during plugin installation

### ISSUE THREAD

**Issue by edgarrmondragon** [meltano/meltano#8625]
### Meltano Version

3.4.2 (and 3.5.0a4 for a new case)

### Python Version

NA

### Bug scope

CLI (options, error messages, logging, etc.)

### Operating System

NA

### Description

An annoying interaction between pipeline-specific environment variables like `MELTANO_EXTRACT__LOAD_SCHEMA` and the `MELTANO_FF_STRICT_ENV_VAR_MODE` feature flag occurs with `meltano install` and #8482.

Consider the following project:

```yaml
version: 1
default_environment: dev
environments:
- name: dev
- name: staging
- name: prod
send_anonymous_usage_stats: false
plugins:
  extractors:
  - name: tap-smoke-test
    variant: meltano
    pip_url: git+https://github.com/meltano/tap-smoke-test.git
    config:
      streams:
      - stream_name: animals
        input_filename: 
          https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
    metadata:
      animals:
        key_properties: [id]
  loaders:
  - name: target-duckdb
    variant: jwills
    pip_url: target-duckdb~=0.6
    config:
      path: output/wh.duckdb
      default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
```

This happens with the latest Meltano (v3.4.2 at the time of writing):

```console
$ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano install
2024-07-12T00:24:23.462889Z [info     ] Installing 2 plugins          
2024-07-12T00:24:23.471755Z [info     ] Installing extractor 'tap-smoke-test'
Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
```

Omitting `MELTANO_FF_STRICT_ENV_VAR_MODE=true` makes the installation succeed, so it's clear the variable `MELTANO_EXTRACT__LOAD_SCHEMA` is expected to have a value, although it's not required in any way to install any of the plugins.

A similar thing now happens in other commands with the introduction of #8482:

```console
$ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano run tap-smoke-test target-duckdb
2024-07-12T00:29:58.442916Z [info     ] Environment 'dev' is active   
2024-07-12T00:29:58.490541Z [info     ] Installing extractor 'tap-smoke-test'
Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
```

The latter can be problematic as it might break users' workflows, so we have a few options if we want to fix this.

### Revert #8482

This is not ideal, since the functionality is useful indeed.

### Make auto-install opt-in

This was actually the original suggestion in https://github.com/meltano/meltano/issues/2089:

> Install operations are not always safe to run, and they can hide other issues in the invocation or installation [...]

This negates a big part of the benefit of the feature, but there'd still a the global toggle in the form of the `auto_install` Meltano setting to turn it on by default.

### Avoid trying to expand values coming from a plugin's config dictionary

This was introduced by https://github.com/meltano/meltano/commit/148a73a69a033397091eace3a912cdd40c14c6f3 and occurs here

https://github.com/meltano/meltano/blob/eb10cdf7309051acb3b1e9c97f7d011a49a13367/src/meltano/core/plugin_install_service.py#L311

and more specifically here

https://github.com/meltano/meltano/blob/eb10cdf7309051acb3b1e9c97f7d011a49a13367/src/meltano/core/plugin_install_service.py#L474

where the plugin's config is expanded. This fails when it refers to a _pipeline_ environment variable such as `MELTANO_EXTRACT__LOAD_SCHEMA`.

This does seem to fix the problem in both scenarios exemplified above, but Chesterton's Fence says that we should preserve this. On the other hand I can't think of a scenario where expanding a plugin's config values would be useful **for installing the plugin**.

### Other workarounds and alternatives

There's of course the option of not using the `strict_env_var_mode` feature flag to shoot yourself in the foot 😅, but there may yet be other workarounds.

cc @WillDaSilva (as the author of 148a73a69a033397091eace3a912cdd40c14c6f3) @ReubenFrankel

---

Related:

* https://github.com/meltano/meltano/issues/2089
* https://github.com/meltano/meltano/issues/7548
* https://github.com/meltano/meltano/pull/7019
* https://github.com/meltano/meltano/pull/8482

**Comment by ReubenFrankel** [comment]
> This does seem to fix the problem in both scenarios exemplified above, but Chesterton's Fence says that we should preserve this. On the other hand I can't think of a scenario where expanding a plugin's config values would be useful for installing the plugin.

Yeah, it's a bit of a difficult one... Would it be possible to differentiate `MELTANO_EXTRACT__LOAD_SCHEMA` as a pipeline variable and ignore it during env var expansion during any command invocation that it would not normally be set under. Is that just `el`/`elt`?

**Comment by edgarrmondragon** [comment]
> Yeah, it's a bit of a difficult one... Would it be possible to differentiate `MELTANO_EXTRACT__LOAD_SCHEMA` as a pipeline variable programmatically and then ignore it during env var expansion for any command invocation that it would not normally be set under? Is that just anything other than `el`/`elt`?

I think ideally and I didn't mention this option above, we could do something similar to what `meltano run` or `meltano el` do under the hood and populate a shared environment with config from all the plugins in the context:

<details><summary><code>meltano run</code></summary>
<p>

https://github.com/meltano/meltano/blob/d8bd1b3cef3c59182f7b9c329753dd5af619d805/src/meltano/core/block/extract_load.py#L237-L260

</p>
</details> 

<details><summary><code>meltano el</code></summary>

<p>

https://github.com/meltano/meltano/blob/d8bd1b3cef3c59182f7b9c329753dd5af619d805/src/meltano/core/elt_context.py#L466-L509

</p>
</details>

**Comment by ReubenFrankel** [comment]
> I think ideally and I didn't mention this option above, we could do something similar to what `meltano run` or `meltano el` do under the hood and populate a shared environment with config from all the plugins in the context:

Specifically for an install? How are you proposing that work for `MELTANO_EXTRACT__LOAD_SCHEMA`?

**Comment by edgarrmondragon** [comment]
> Specifically for an install?

Yeah, we'd pass all the merged env dictionary from all plugin contexts to each individual installation.

> How are you proposing that work for `MELTANO_EXTRACT__LOAD_SCHEMA`?

I think the problem is that `MELTANO_EXTRACT__LOAD_SCHEMA` gets populated for the extractor, but it is required for the loader and not populated independently there. So in fact this works fine already:

```
$ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano invoke tap-smoke-test
024-07-12T17:16:13.787617Z [info     ] Environment 'dev' is active   
2024-07-12T17:16:13.823082Z [info     ] Installing extractor 'tap-smoke-test'
2024-07-12T17:16:25.220142Z [info     ] Installed extractor 'tap-smoke-test'
2024-07-12 11:16:28,863 | INFO     | root                 | animals stream max schema_inference_record_count hit
{"type":"STATE","value":{}}
2024-07-12 11:16:28,864 | INFO     | tap-smoke-test.animals | Beginning full_table sync of 'animals'...
2024-07-12 11:16:28,864 | INFO     | tap-smoke-test.animals | Tap has custom mapper. Using 1 provided map(s).
...
```

**Comment by ReubenFrankel** [comment]
I'm finding this a bit hard to follow, but are there not two separate issue here:

1. Plugin contexts are not shared at install time
    ```sh
    env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    ```

    <details>
    <summary>Working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```
    <details>
    <summary>Not working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
          env:
            TEST: $MELTANO_EXTRACT__LOAD_SCHEMA
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:19:33.171147Z [info     ] Environment 'dev' is active   
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ```

1. `MELTANO_EXTRACT__LOAD_SCHEMA` is not available for certain commands (maybe others variables as well)

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    <details>
    <summary><code>el</code>/<code>elt</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```

    </details>

    <details>
    <summary><code>run</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano run tap-smoke-test target-duckdb
    2024-07-15T01:29:41.990516Z [info     ] Environment 'dev' is active   
    2024-07-15T01:29:42.133819Z [info     ] Installing extractor 'tap-smoke-test'
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ...
    ```

    </details>

    This is not just an issue for `run`, but all commands that support auto-install other than `el`/`elt`. Not sure what the best solution is for this in particular, but

    > ### Avoid trying to expand values coming from a plugin's config dictionary

    seems like it could work,

**Comment by ReubenFrankel** [comment]
I'm finding this a bit hard to follow, but are there not two separate issues here:

1. Plugin contexts are not shared at install time:
    ```sh
    env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    ```

    <details>
    <summary>Working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```
    <details>
    <summary>Not working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
          env:
            TEST: $MELTANO_EXTRACT__LOAD_SCHEMA
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:19:33.171147Z [info     ] Environment 'dev' is active   
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ```

1. `MELTANO_EXTRACT__LOAD_SCHEMA` is not available for certain commands (maybe others variables as well)

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    <details>
    <summary><code>el</code>/<code>elt</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```

    </details>

    <details>
    <summary><code>run</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano run tap-smoke-test target-duckdb
    2024-07-15T01:29:41.990516Z [info     ] Environment 'dev' is active   
    2024-07-15T01:29:42.133819Z [info     ] Installing extractor 'tap-smoke-test'
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ...
    ```

    </details>

    This is not just an issue for `run`, but all commands that support auto-install other than `el`/`elt`. Not sure what the best solution is for this in particular, but

    > ### Avoid trying to expand values coming from a plugin's config dictionary

    seems like it could work,

**Comment by ReubenFrankel** [comment]
I'm finding this a bit hard to follow, but are there not two separate issues here:

1. Plugin contexts are not shared at install time
    ```sh
    env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    ```

    <details>
    <summary>Working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```
    <details>
    <summary>Not working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
          env:
            TEST: $MELTANO_EXTRACT__LOAD_SCHEMA
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:19:33.171147Z [info     ] Environment 'dev' is active   
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ```

1. `MELTANO_EXTRACT__LOAD_SCHEMA` is not available for certain commands (maybe others variables as well)

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    <details>
    <summary><code>el</code>/<code>elt</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```

    </details>

    <details>
    <summary><code>run</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano run tap-smoke-test target-duckdb
    2024-07-15T01:29:41.990516Z [info     ] Environment 'dev' is active   
    2024-07-15T01:29:42.133819Z [info     ] Installing extractor 'tap-smoke-test'
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ...
    ```

    </details>

    This is not just an issue for `run`, but all commands that support auto-install other than `el`/`elt`. Not sure what the best solution is for this in particular, but

    > ### Avoid trying to expand values coming from a plugin's config dictionary

    seems like it could work,

**Comment by ReubenFrankel** [comment]
I'm finding this a bit hard to follow, but are there not two separate issues here:

1. Plugin contexts are not shared at install time
    ```sh
    env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    ```

    <details>
    <summary>Working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```
    <details>
    <summary>Not working</summary>

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
          env:
            TEST: $MELTANO_EXTRACT__LOAD_SCHEMA
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    </details>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:19:33.171147Z [info     ] Environment 'dev' is active   
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ```

1. `MELTANO_EXTRACT__LOAD_SCHEMA` is not available for certain commands (maybe other variables as well)

    ```yml
    version: 1
    default_environment: dev
    environments:
    - name: dev
    - name: staging
    - name: prod
    send_anonymous_usage_stats: false
    plugins:
      extractors:
      - name: tap-smoke-test
        variant: meltano
        pip_url: git+https://github.com/meltano/tap-smoke-test.git
        config:
          streams:
          - stream_name: animals
            input_filename: 
              https://raw.githubusercontent.com/meltano/tap-smoke-test/main/demo-data/animals-data.jsonl
        metadata:
          animals:
            key_properties: [id]
      loaders:
      - name: target-duckdb
        variant: jwills
        pip_url: target-duckdb~=0.6
        config:
          path: output/wh.duckdb
          default_target_schema: $MELTANO_EXTRACT__LOAD_SCHEMA
    ```

    <details>
    <summary><code>el</code>/<code>elt</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano el tap-smoke-test target-duckdb
    2024-07-15T01:20:45.755049Z [info     ] Environment 'dev' is active   
    2024-07-15T01:20:45.969319Z [info     ] Installing extractor 'tap-smoke-test'
    2024-07-15T01:20:46.010596Z [info     ] Installing loader 'target-duckdb'
    ...
    ```

    </details>

    <details>
    <summary><code>run</code></summary>

    ```sh
    $ env -i PATH="$PATH" HOME=$HOME MELTANO_FF_STRICT_ENV_VAR_MODE=true meltano run tap-smoke-test target-duckdb
    2024-07-15T01:29:41.990516Z [info     ] Environment 'dev' is active   
    2024-07-15T01:29:42.133819Z [info     ] Installing extractor 'tap-smoke-test'
    Environment variable 'MELTANO_EXTRACT__LOAD_SCHEMA' referenced but not set. Make sure the environment variable is set.
    ...
    ```

    </details>

    This is not just an issue for `run`, but all commands that support auto-install other than `el`/`elt`. Not sure what the best solution is for this in particular, but

    > ### Avoid trying to expand values coming from a plugin's config dictionary

    seems like it could work,

**Comment by edgarrmondragon** [comment]
> 2\. > ### Avoid trying to expand values coming from a plugin's config dictionary
>         
>         
>         seems like it could work.

Seems like the easiest option. Let's actually wait and see if someone complains 😅.

I don't think anyone complained about https://github.com/meltano/meltano/issues/7548 but me, so this might be the case again.

**Comment by edgarrmondragon** [comment]
In the meantime, https://github.com/meltano/meltano/pull/8633 might help folks do something like:

```
# Only install the required plugins, don't run the pipeline
meltano run --only-install tap-hubspot target-postgres

# Don't install the already installed required plugins, only run the pipeline
meltano run --no-install tap-hubspot target-postgres
```
